core/src/main/java/io/grpc/internal/ServiceConfigState.java

```diff
@@ -0,0 +1,135 @@
+import static com.google.common.base.Preconditions.checkState;
+
+import io.grpc.Status;
+import io.grpc.SynchronizationContext;
+import javax.annotation.Nullable;
+
+/**
+ * {@link ServiceConfigState} holds the state of the current service config.  It must be mutated
+ * and read from {@link ManagedChannelImpl} constructor or the provided syncContext.
+ */
+final class ServiceConfigState {
+  @Nullable private final ManagedChannelServiceConfig defaultServiceConfig;
+  private final boolean lookUpServiceConfig;
+  private final SynchronizationContext syncCtx;
+
+  // mutable state
+  @Nullable private ManagedChannelServiceConfig currentServiceConfig;
+  @Nullable private Status currentError;
+
+  /**
+   * @param defaultServiceConfig The initial service config, or {@code null} if absent.
+   * @param lookUpServiceConfig {@code true} if service config updates might occur.
+   * @param syncCtx The synchronization context that this is accessed from.
+   */
+  ServiceConfigState(
+      @Nullable ManagedChannelServiceConfig defaultServiceConfig,
+      boolean lookUpServiceConfig,
+      SynchronizationContext syncCtx) {
+    this.defaultServiceConfig = defaultServiceConfig;
+    this.lookUpServiceConfig = lookUpServiceConfig;
+    this.syncCtx = checkNotNull(syncCtx, "syncCtx");
+    if (!lookUpServiceConfig) {
+      this.currentServiceConfig = defaultServiceConfig;
+    }
+  }
+
+  /**
+   * Returns {@code true} if it RPCs should wait on a service config resolution.  This can return
+   * {@code false} if:
+   *
+   * <ul>
+   *   <li>There is a valid service config from the name resolver
+   *   <li>There is a valid default service config and a service config error from the name
+   *       resolver
+   *   <li>No service config from the name resolver, and no intent to lookup a service config.
+   *
+   * In the final case, the default service config may be present or absent, and will be the
+   * current service config.
+   */
+  boolean waitOnServiceConfig() {
+    syncCtx.throwIfNotInThisSynchronizationContext();
+    if (currentServiceConfig != null || currentError != null) {
+      return false;
+    } else {
+      return expectUpdates();
+    }
+  }
+
+  /**
+   * Gets the current service config.
+   *
+   * @throws IllegalStateException if the service config has not yet been updated.
+   */
+  @Nullable ManagedChannelServiceConfig getCurrentServiceConfig() {
+    checkState(!waitOnServiceConfig(), "still waiting on service config");
+    return currentServiceConfig;
+  }
+
+  /**
+   * Gets the current service config error.
+   *
+   * @throws IllegalStateException if the service config has not yet been updated.
+   */
+  @Nullable Status getCurrentError() {
+    checkState(!waitOnServiceConfig(), "still waiting on service config");
+    return currentError;
+  }
+
+  /**
+   * Returns {@code true} if the update was successfully applied, else {@code false}.
+   */
+  boolean update(Status error) {
```

Comment:
I plan on either logging it if the return value is false, or not and letting the channel go into transient failure.

Reply:
Theres only two possible ways this method can work:

1.  The error is installed 
2.  The error is ignored.

If the error is installed into this state, the ManagedChannel can enter transient failure as described in the gRFC.    If it is not installed, it is notable to log.  The return value of this indicates if the installation was successful.  The Channel is the actual consumer of this error, not the ConfigServiceState.  It would be a violation of responsibility to handle the error here.

Reply:
I don't necessarily agree with you about the split of responsibilities, but it's minor.